### PR TITLE
add farthest point sampling for points sampler, and modify the uniform sampler for the meshes

### PR DESCRIPTION
**Pull request by hansen7**
Hi, Castro

I have added farthest point sampling into the points sampler, which is an algorithm applied to the cutting-edge research, see [here](https://github.com/charlesq34/pointnet) for the relevant paper, and [here](https://flothesof.github.io/farthest-neighbors.html) for the description of the algorithm.

And I think it is probably better to sample `u, v` uniformly from `[0, 1]` and `[0, u]` respectively in the mesh sampler. Because to ensure the sampled points are within that face, it requires not only `u + v  <= 1` but also `u >= 0` and `v >= 0`, but for the original procedure, which is sampled from Gaussian then truncation, it did not resolve the issues with negative sampling results. Plus it is more concise now.

I am very interested to make more commitments since it intersects with my research projects during PhD, I think next step is to develop the Poisson Disk Sampling as discussed here: #239, and is it okay for me to develop a new part for the normals/curvatures calculations(directly from the points) and with reconstruction on the doc as well.

Thanks.

### CODE REVIEW

**Comment by daavoo** [comment]
Hola @hansen7,

Thank you so much for the contribution! I'm leaving a few comments in the P.R. review.

Pd. I'm glad to hear that you are interested in making more commitments, don't hesitate to ask anything.

File: pyntcloud/samplers/points.py [hunk 1]
```diff
@@ -24,3 +27,42 @@ def compute(self):
...
+            raise ValueError("the Mahalanobis matrix must be positive semidefinite")
+        self.d_metric = d_metric
+
+    def cal_distance(self, point, solution_set):
+        sum = 0
+        for pt in solution_set:
+            sum += np.dot(point[:3]-pt[:3], self.d_metric)
+        return sum
+
+    def compute(self):
```
Comment (daavoo) [inline pyntcloud/samplers/points.py]: I think that this implementation was originally though for working with python lists. However PyntCloud is using pandas DataFrames/Numpy arrays internally.

This has 2 main implications:

- The current code won't work as it is right now due to the usage of `remaining_points.pop`, which is causing the following error:

```python
AttributeError: 'numpy.ndarray' object has no attribute 'pop'
```

- The code could be rewritten more efficient and concise using one or more of the following options:

a) numpy vectorization:
https://realpython.com/numpy-array-programming/
b) numba
https://numba.pydata.org/numba-doc/dev/user/5minguide.html
c) pandas operations
https://pandas.pydata.org/pandas-docs/stable/getting_started/basics.html

**Comment by daavoo** [comment]
In addition to the above comments, please follow the steps described in:

https://pyntcloud.readthedocs.io/en/latest/samplers_dev.html#let-pyntcloud-know-about-your-sampler

In order to make your new sample available and easily accesible for everybody

File: pyntcloud/samplers/mesh.py [hunk 2]
```diff
@@ -71,12 +71,8 @@ def compute(self):
...
 
         # (n, 1) the 1 is for broadcasting
-        u = np.random.rand(self.n, 1)
-        v = np.random.rand(self.n, 1)
-        is_a_problem = u + v > 1
-
-        u[is_a_problem] = 1 - u[is_a_problem]
-        v[is_a_problem] = 1 - v[is_a_problem]
+        u = np.random.uniform(low=0., high=1., size=(self.n, 1))
+        v = np.random.uniform(low=0., high=u, size=(self.n, 1))
```
[the remark this answers is not in the document]
Comment (hansen7) [reply pyntcloud/samplers/mesh.py]: oh, yes, sorry for the typos, I was meant to write `1-u`

File: pyntcloud/io/obj.py [hunk 3]
```diff
@@ -72,7 +72,7 @@ def read_obj(filename):
         for i in range(sum(c.isdigit() for c in f[0].split(" "))):
             mesh_columns.append("v{}".format(i + 1))
 
-    mesh = pd.DataFrame([re.split(r'\D+', x) for x in f], dtype='i4', columns=mesh_columns)
+    mesh = pd.DataFrame([re.split(r'\D+', x) for x in f], dtype='i4', columns=mesh_columns).astype('i4')
```
[the remark this answers is not in the document]
Comment (hansen7) [reply pyntcloud/io/obj.py]: no worries :)

**Comment by hansen7** [comment]
> 

haha no worries, is there other places for development, I am looking forward to more contributions

**Pull request merged**